docs/source/notes/modules.rst

```diff
@@ -0,0 +1,433 @@
+     net.zero_grad()
+     loss.backward()
+     optimizer.step()
+
+In this simplified example, the network simply learns to output zero, which is not very interesting, but the
+key parts of training are present:
+
+* A network is created.
+* An optimizer (in this case, a stochastic gradient descent optimizer) is created, and the network’s
+  parameters are associated with it.
+* A training loop...
+    * acquires an input,
+    * runs the network,
+    * computes a loss,
+    * zeros the network’s parameters’ gradients,
+    * calls loss.backward() to update the parameters’ gradients,
+    * calls optimizer.step() to apply the gradients to the parameters.
+
+After the above snippet has been run, note that the network's parameters have changed. In particular, examining the
+value of ``l1``\ 's ``weight`` parameter shows that its values are now much closer to 0 (as may be expected):
+
+.. code-block:: python
+
+   print(net.l1.weight)
+   : Parameter containing:
+   tensor([[-0.0013],
+           [ 0.0030],
+           [-0.0008]], requires_grad=True)
+
+Training neural networks can often be tricky. For more information, check out:
+
+* Using optimizers: https://pytorch.org/tutorials/beginner/examples_nn/two_layer_net_optim.html.
+* Neural network training: https://pytorch.org/tutorials/beginner/blitz/neural_networks_tutorial.html
+* Introduction to autograd: https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html
+
+Module State
+------------
+
+In the previous section, we demonstrated training a module's "parameters", or learnable aspects of computation.
+Now, if we want to save out the trained model, we can do so by saving its ``state_dict`` (i.e. "state dictionary"):
+
+.. code-block:: python
+
+   # Save the module
+   torch.save(net.state_dict(), 'net.pt')
+
+   ...
+
+   # Load the module later on
+   new_net = Net()
+   new_net.load_state_dict(torch.load('net.pt'))
+   : <All keys matched successfully>
+
+A module's ``state_dict`` contains state that affects its computation. This includes, but is not limited to, the
+module's parameters. For some modules, it may be useful to have state beyond parameters that affects module
+computation but is not learnable. For such cases, PyTorch provides the concept of "buffers", both "persistent"
+and "non-persistent". Following is an overview of the various types of state a module can have:
+
+* **Parameters**\ : learnable aspects of computation; contained within the ``state_dict``
+* **Buffers**\ : non-learnable aspects of computation
+
+  * **Persistent** buffers: contained within the ``state_dict`` (i.e. serialized when saving)
+  * **Non-persistent** buffers: not contained within the ``state_dict`` (i.e. left out of serialization)
+
+As a motivating example for the use of buffers, consider a simple module that maintains a running mean. We want
+the current value of the running mean to be considered part of the module's ``state_dict`` and serialized, but not
```

Comment:
why do we want it to be part of the module's state_dict?  (Aside: this reasoning seems backwards from how I would expect users to think about it, i.e. they think about what they want restored after loading, therefore it should be in the state_dict).

Reply:
Changed the wording to focus on restoration during loading.